chain/consensus/mir/manager.go

```diff
@@ -0,0 +1,355 @@
+
+	logger := newManagerLogger()
+
+	// Create Mir modules.
+	// TODO: Configure repo path to persist Mir WAL
+	path := fmt.Sprintf("lotus-wal%s", strings.Replace(mirID, "/", "-", -1))
+	wal, err := NewWAL(mirID, path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create WAL: %w", err)
+	}
+
+	netTransport, err := mirlibp2p.NewTransport(h, t.NodeID(mirID), logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create transport: %w", err)
+	}
+	if err := netTransport.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start transport: %w", err)
+	}
+	netTransport.Connect(ctx, initialMembership)
+
+	cryptoManager, err := NewCryptoManager(addr, api)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create crypto manager: %w", err)
+	}
+
+	// Instantiate an interceptor.
+	var interceptor *eventlog.Recorder
+
+	m := Manager{
+		Addr:                addr,
+		NetName:             netName,
+		Pool:                fifo.New(),
+		MirID:               mirID,
+		interceptor:         interceptor,
+		WAL:                 wal,
+		CryptoManager:       cryptoManager,
+		Net:                 netTransport,
+		InitialValidatorSet: initialValidatorSet,
+		ToMir:               make(chan chan []*mirproto.Request),
+	}
+
+	m.StateManager = NewStateManager(initialMembership, &m)
+
+	mpool := pool.NewModule(
+		m.ToMir,
+		&pool.ModuleConfig{
+			Self:   "mempool",
+			Hasher: "hasher",
+		},
+		&pool.ModuleParams{
+			MaxTransactionsInBatch: 10,
+		},
+	)
+
+	smrSystem, err := smr.New(
+		t.NodeID(mirID),
+		h,
+		initialMembership,
+		m.CryptoManager,
+		m.StateManager,
+		logger,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not create SMR system: %w", err)
+	}
+	smrSystem = smrSystem.WithModule("mempool", mpool)
+
+	if err := smrSystem.Start(ctx); err != nil {
+		return nil, fmt.Errorf("could not start SMR system: %w", err)
+	}
+
+	cfg := mir.DefaultNodeConfig().WithLogger(logger)
+
+	interceptorOutput := os.Getenv(InterceptorOutputEnv)
+	if interceptorOutput != "" {
+		// TODO: Persist in repo path?
+		m.interceptor, err = eventlog.NewRecorder(
+			t.NodeID(mirID),
+			interceptorOutput,
+			logging.Decorate(logger, "Interceptor: "),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create interceptor: %w", err)
+		}
+		m.MirNode, err = mir.NewNode(t.NodeID(mirID), cfg, smrSystem.Modules(), wal, m.interceptor)
+	} else {
+		m.MirNode, err = mir.NewNode(t.NodeID(mirID), cfg, smrSystem.Modules(), wal, nil)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Mir node: %w", err)
+	}
+
+	return &m, nil
+}
+
+// Start starts the manager.
+func (m *Manager) Start(ctx context.Context) chan error {
+	log.Infof("Mir manager %s starting", m.MirID)
+
+	errChan := make(chan error, 1)
+
+	go func() {
+		// Run Mir node until it stops.
+		if err := m.MirNode.Run(ctx); err != nil && !errors.Is(err, mir.ErrStopped) {
+			log.Infof("Mir manager %s: Mir node stopped with error: %v", m.MirID, err)
+			errChan <- err
+		}
```

Comment:
In the current implementation, this will lead to panicking on stop due to [this](https://github.com/consensus-shipyard/lotus/blob/adlrocha/mir/chain/consensus/mir/mine.go#L73-L78).

Reply:
I see... Would it make sense to check for nil on the receiving end and only panic on non-nil errors?
Receiving a nil error might be a good indication to even return from the `Mine` function...

Reply:
Fixed